### DISCUSSION TITLE
v0.14.0

### DISCUSSION THREAD

**Discussion by laipz8200** [langgenius/dify#11692] [Releases]
# ✨ Discover What's New in v0.14.0 ✨

We’re excited to bring you the latest v0.14.0 update. Let's dive straight into some of the standout features and improvements. As always, we extend a special thanks to our contributors!

### 🔗 **Improved Error Handling**

![continue on error](https://github.com/user-attachments/assets/28195653-9d35-4455-83ec-42b6eca107cf)

Enhanced functionality for workflows and chatflows allows nodes like LLM and HTTP requests to continue operations even when errors occur. Big shoutout to @zxhlyh and @Nov1c444 for their work in [#11474](https://github.com/langgenius/dify/pull/11474) and [#11458](https://github.com/langgenius/dify/pull/11458).

### 🎨 **Revamped User Interface**

| Creation | Exploration |
|--------|-------|
| ![create](https://github.com/user-attachments/assets/e112db5b-1c78-46da-ad79-1ec8e7c567cf)    | ![template](https://github.com/user-attachments/assets/a0dbf723-d0e9-4836-943f-3a032e9991bc)   |

With the fresh entry point for app creation and template exploration, user experience becomes more intuitive, assisting in quick and efficient startup. Thanks to @douxc for this in [#10847](https://github.com/langgenius/dify/pull/10847).

### 📁 **File Previews Directly in Chat**

Now, video, audio, and PDF file previews are accessible right within the chat, simplifying on-the-go information access. Credit goes to @charli117 in [#11466](https://github.com/langgenius/dify/pull/11466).

Learn more in [our blog](https://dify.ai/blog/boost-ai-workflow-resilience-with-error-handling)

### 🌍 **Gemini 2.0 and Flash Exp in Google**
This enhancement by @hjlarry and @wangiii, found in [#11570](https://github.com/langgenius/dify/pull/11570) and [#11604](https://github.com/langgenius/dify/pull/11604), boosts overall performance.

### 🔢 **Grok-2-1212 and Grok-2-Vision-1212 in XAI**
Optimizing the model functionality, thanks to @hjlarry in [#11672](https://github.com/langgenius/dify/pull/11672).

### 🧙 **QwQ and Llama3.3 Models in Siliconflow**
Expands compatibility and versatility, contributed by @orangeclk in [#11492](https://github.com/langgenius/dify/pull/11492).

### 📷 **Abab6.5t-Chat Model in Minimax**
Broaden available models by @acelyc111 in [#11484](https://github.com/langgenius/dify/pull/11484).

### 🏦 **Storage Integration**
Features integration of [Apache OpenDAL](https://github.com/apache/opendal) storage for enhanced system-wide storage capabilities. Work done by @laipz8200 in [#11508](https://github.com/langgenius/dify/pull/11508).

### 📖 **Knowledge Base Enhancements**
Added support for MDX files in the Knowledge Base to improve content flexibility. Credit to @yihong0618 in [#11565](https://github.com/langgenius/dify/pull/11565).

### 🔧 **Logging and Dev Tools**
Logs now include trace IDs for smoother debugging, thanks to @luckylhb90 in [#11599](https://github.com/langgenius/dify/pull/11599). Plus, @fujita-h added handy Docker commands in [#11616](https://github.com/langgenius/dify/pull/11616).

### 🔧 **Bug Fixes**
Various minor bugs have been addressed, including iteration node errors and proper release of rate-restricting requests. Thanks to @yihong0618, @liuzhenghua, and others for resolving these issues in [#11504](https://github.com/langgenius/dify/pull/11504), [#11540](https://github.com/langgenius/dify/pull/11540), and more.

### 🌈 **Additional Enhancements**
Lindorm VDB support and configurable max submit count, thanks to @AlwaysBluer and @kazuhisa-wada in [#11574](https://github.com/langgenius/dify/pull/11574) and [#11673](https://github.com/langgenius/dify/pull/11673).

Every update in this release reinforces the platform's robustness and adaptability—many thanks to all contributors! Let's continue building this powerful community together! 🚀

---

### Docker compose deployments

> [!WARNING]
>
> The files in the docker-legacy directory will soon stop being maintained and will be removed from the repository. If you are still using them, please switch to the new version as soon as possible.

1. Back up your customized docker-compose YAML file (optional)

   ```bash
   cd docker
   cp docker-compose.yaml docker-compose.yaml.$(date +%s).bak
   ```

2. Get the latest code from the main branch

   ```bash
   git checkout main
   git pull origin main
   ```

3. Stop the service，Command, please execute in the docker directory

   ```bash
   docker compose down
   ```

4. Back up data

   ```bash
   tar -cvf volumes-$(date +%s).tgz volumes
   ```

5. Upgrade services

   ```bash
   docker compose up -d
   ```

### Source Code deployments

1. Stop API server, Worker and Web frontend Server.

2. Get the latest code from the release branch:

   ```bash
   git checkout 0.14.0
   ```

3. Update Python dependencies:

   ```bash
   cd api
   poetry install
   ```

7. Then, let's run the migration script:

   ```bash
   poetry run flask db upgrade
   ```

8. Finally, run API server, Worker and Web frontend Server again.

---

## What's Changed
* fix: add the missing abab6.5t-chat model of Minimax by @acelyc111 in https://github.com/langgenius/dify/pull/11484
* feat:add apollo configuration to load env file by @huanshare in https://github.com/langgenius/dify/pull/11210
* feat: add siliconflow qwq and llama3.3 model by @orangeclk in https://github.com/langgenius/dify/pull/11492
* fix(app_generator_service): overload type hints by @laipz8200 in https://github.com/langgenius/dify/pull/11507
* chore : fix translation Typo in ko-KR localization by @JungHyunMoon in https://github.com/langgenius/dify/pull/11509
* chore: fix typo in Japanese localization by @barabicu in https://github.com/langgenius/dify/pull/11502
* Msg file preview by @charli117 in https://github.com/langgenius/dify/pull/11466
* chore: LOCAL_FILE also try to use remote_url as Prompt message by @hjlarry in https://github.com/langgenius/dify/pull/11443
* Remove the processing of single quote when testing API tools. by @weimzh in https://github.com/langgenius/dify/pull/11390
* fix: issue 11247 that Completion mode content maybe list or str by @yihong0618 in https://github.com/langgenius/dify/pull/11504
* fix: can not start local by REMOTE_SETTINGS_SOURCE_NAME change it to … by @yihong0618 in https://github.com/langgenius/dify/pull/11535
* Revert "fix: total tokens is wrong which is zero in inter way, close … by @yihong0618 in https://github.com/langgenius/dify/pull/11536
* fix: better error message for url add external knowledge by @yihong0618 in https://github.com/langgenius/dify/pull/11537
* feat: Add llama-3.3 models for Groq by @pvoo in https://github.com/langgenius/dify/pull/11533
* fix: Remove duplicate 'response_format' parameter from model YAML files by @zkytech in https://github.com/langgenius/dify/pull/11531
* [Pixtral] Add new model ; add vision by @Asterovim in https://github.com/langgenius/dify/pull/11231
* chore: update thai lang in app page by @YIXIAO0 in https://github.com/langgenius/dify/pull/11541
* feat: workflow continue on error by @zxhlyh in https://github.com/langgenius/dify/pull/11474
* Feat: continue on error by @Nov1c444 in https://github.com/langgenius/dify/pull/11458
* fix: iteration node in parallel mode token count error by @Nov1c444 in https://github.com/langgenius/dify/pull/11539
* feat: integrate opendal storage by @laipz8200 in https://github.com/langgenius/dify/pull/11508
* fix: workflow continue on error doc link by @zxhlyh in https://github.com/langgenius/dify/pull/11554
* chore: translate i18n files by @github-actions in https://github.com/langgenius/dify/pull/11545
* Fix: RateLimit requests were not released when a streaming generation exception occurred by @liuzhenghua in https://github.com/langgenius/dify/pull/11540
* fix: better opendal tests by @yihong0618 in https://github.com/langgenius/dify/pull/11569
* feat: add gemini-2.0-flash-exp by @hjlarry in https://github.com/langgenius/dify/pull/11570
* Lindorm vdb by @AlwaysBluer in https://github.com/langgenius/dify/pull/11574
* Feat: dark mode for logs and annotations by @JzoNgKVO in https://github.com/langgenius/dify/pull/11575
* fix: change workflow trace id by @ZhouhaoJiang in https://github.com/langgenius/dify/pull/11585
* fix: support mdx files close #11557 by @yihong0618 in https://github.com/langgenius/dify/pull/11565
* chore: translate i18n files by @github-actions in https://github.com/langgenius/dify/pull/11577
* [ref] use one method to get boto client for aws bedrock by @warren830 in https://github.com/langgenius/dify/pull/11506
* Added new models and Removed the deleted ones for Groq #11455 by @alokas in https://github.com/langgenius/dify/pull/11456
* fix: name of llama-3.3-70b-specdec by @crazywoola in https://github.com/langgenius/dify/pull/11596
* feat(model): add vertex_ai Gemini 2.0 Flash Exp by @wangiii in https://github.com/langgenius/dify/pull/11604
* ci: better print version for ruff to check the change by @yihong0618 in https://github.com/langgenius/dify/pull/11587
* feat(devcontainer): add alias to stop Docker containers by @fujita-h in https://github.com/langgenius/dify/pull/11616
* Fix/pdf preview in build by @zxhlyh in https://github.com/langgenius/dify/pull/11621
* chore(dependency): bump gunicorn to 23.0 by @laipz8200 in https://github.com/langgenius/dify/pull/11560
* fix: split dir for opendal tests by @yihong0618 in https://github.com/langgenius/dify/pull/11627
* fix: app log filter value error by @Kevin9703 in https://github.com/langgenius/dify/pull/11624
* Feat: new entry point for app creation by @douxc in https://github.com/langgenius/dify/pull/10847
* fix: filter bug for keywork cause code can not reach by @yihong0618 in https://github.com/langgenius/dify/pull/11666
* fix: remove unnecessary curly braces in wf api doc by @RockChinQ in https://github.com/langgenius/dify/pull/11658
* fix: rename README filename: Japanese language code is 'JA' by @iwaim in https://github.com/langgenius/dify/pull/11651
* fix: langfuse do not have created_at args and fix the typing in the file by @yihong0618 in https://github.com/langgenius/dify/pull/11648
* Update translate to Japanese: natural Japanese expression by @iwaim in https://github.com/langgenius/dify/pull/11647
* fix: better error message for stream by @yihong0618 in https://github.com/langgenius/dify/pull/11635
* test: run vdb tests on TiDB Vector with docker in CI tests by @bowenliang123 in https://github.com/langgenius/dify/pull/11645
* fix: account.id should account_id by @yihong0618 in https://github.com/langgenius/dify/pull/11628
* fix: ExternalDatasetService.process_external_api wrong args by @yihong0618 in https://github.com/langgenius/dify/pull/11586
* fix: _handle_workflow_run_partial_success args is wrong by @yihong0618 in https://github.com/langgenius/dify/pull/11562
* feat: log add trace id by @luckylhb90 in https://github.com/langgenius/dify/pull/11599
* chore: translate i18n files by @github-actions in https://github.com/langgenius/dify/pull/11639
* fix: remove the unused QueueWorkflowPartialSuccessEvent handle in workflow by @Nov1c444 in https://github.com/langgenius/dify/pull/11669
* feat: add grok-2-1212 and grok-2-vision-1212 by @hjlarry in https://github.com/langgenius/dify/pull/11672
* fix: change http node params from dict to list tuple by @hjlarry in https://github.com/langgenius/dify/pull/11665
* fix: deepseek reports an error when using Response Format #11677 by @zhaobingshuang in https://github.com/langgenius/dify/pull/11678
* Make max_submit_count configurable via Config by @kazuhisa-wada in https://github.com/langgenius/dify/pull/11673
* fix: volcengine_maas and baichuan message error by @zhongliliu-butterfly in https://github.com/langgenius/dify/pull/11625
* fix: enhance workflow.tool_published performance by @wangbin77 in https://github.com/langgenius/dify/pull/11640
* fix: tool constant params change cause page crashed by @iamjoel in https://github.com/langgenius/dify/pull/11682
* fix: iteration output array type causes always outputting string array by @iamjoel in https://github.com/langgenius/dify/pull/11686
* fix: workflow continue on error edge color by @zxhlyh in https://github.com/langgenius/dify/pull/11689
* chore: bump version to 0.14.0 by @laipz8200 in https://github.com/langgenius/dify/pull/11679

## New Contributors
* @acelyc111 made their first contribution in https://github.com/langgenius/dify/pull/11484
* @huanshare made their first contribution in https://github.com/langgenius/dify/pull/11210
* @JungHyunMoon made their first contribution in https://github.com/langgenius/dify/pull/11509
* @barabicu made their first contribution in https://github.com/langgenius/dify/pull/11502
* @weimzh made their first contribution in https://github.com/langgenius/dify/pull/11390
* @pvoo made their first contribution in https://github.com/langgenius/dify/pull/11533
* @zkytech made their first contribution in https://github.com/langgenius/dify/pull/11531
* @Asterovim made their first contribution in https://github.com/langgenius/dify/pull/11231
* @alokas made their first contribution in https://github.com/langgenius/dify/pull/11456
* @iwaim made their first contribution in https://github.com/langgenius/dify/pull/11651
* @zhongliliu-butterfly made their first contribution in https://github.com/langgenius/dify/pull/11625
* @wangbin77 made their first contribution in https://github.com/langgenius/dify/pull/11640

**Full Changelog**: https://github.com/langgenius/dify/compare/0.13.2...0.14.0

<hr /><em>This discussion was created from the release <a href='https://github.com/langgenius/dify/releases/tag/0.14.0'>v0.14.0</a>.</em>

**Comment by breakstring** [comment]
## OOM Exception after upgraded to 0.14.0

After upgrading from version 0.13.2 to 0.14.0 using Docker-Compose, the service crashed. I was even unable to SSH into my remote server (Azure VM). After several reboots, I still couldn't connect, but later saw the following message from the serial console of the VM in Azure:

dify login: [   27.033521] cloud-init[2474]: Cloud-init v. 24.3.1-0ubuntu0~24.04.2 running 'modules:final' at Mon, 16 Dec 2024 09:24:29 +0000. Up 26.99 seconds.
[   27.059871] cloud-init[2474]: Cloud-init v. 24.3.1-0ubuntu0~24.04.2 finished at Mon, 16 Dec 2024 09:24:29 +0000. Datasource DataSourceAzure [seed=/var/lib/waagent].  Up 27.05 seconds
[  134.663354] Out of memory: Killed process 2249 (celery) total-vm:35212808kB, anon-rss:736240kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:2220kB oom_score_adj:0

dify login: [ 2127.518524] Out of memory: Killed process 21467 (gunicorn) total-vm:35131260kB, anon-rss:727544kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:2200kB oom_score_adj:0

It seems that the programs related to these OOM error messages are all associated with Dify.
I tried upgrading the server configuration from 4C8G to 4C16G, and now I can finally log in to the server normally. From the following commands, we can also see that the API service is now taking up quite a bit of memory. It seems that running on 4C8G would be quite challenging.

![image](https://github.com/user-attachments/assets/bc5d6b34-b404-47b4-9959-a05be49591f6)

**Reply by laipz8200** [reply]
If you modified SERVER_WORKER_AMOUNT, try to keep it set to 1. Normally, `docker-api` will use 900 - 1500MB of memory.

**Reply by breakstring** [reply]
> If you modified SERVER_WORKER_AMOUNT, try to keep it set to 1. Normally, `docker-api` will use 900 - 1500MB of memory.

Oops, I guess I might have changed this parameter. 
I remember the .env.example had a description suggesting adjusting it according to the number of CPU cores. However, I forgot what I changed it to. 
It seems necessary to remind that it should also be set according to the memory size in the example file.

Thank you so much.

**Reply by laipz8200** [reply]
Thank you for your reminder, I'll update the suggestion. For Gevent, it should always be 1.

**Reply by bowenliang123** [reply]
Shall we pin SERVER_WORKER_AMOUNT to 1 for API service?

**Reply by laipz8200** [reply]
@bowenliang123 We also allow users to use workers other than Gevent. This can be suggested as an option and does not need to be pinned.

**Reply by bowenliang123** [reply]
Sure, sgtm.
